API endpoint for refreshing watchlist instances

xinyis991105 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
This PR adds the API endpoint for bulk refreshing watchlist instances for a course. In addition, it also patched app/models/risk_conditions.rb such that the method update_current_for_course now utilizes transaction.

## Motivation and Context
This is part of the metrics feature functionality.

## How Has This Been Tested?
1. Add a new course with a reasonable number of grace days.
2. Add at least 3 users to that course.
3. Add a reasonable number of assignments, each consuming preferably different number of grace days (by tweaking due date and end date). For ease of later testing, set only one problem to each assignment with weight 100. Act as each user added in step 2. Submit as these users such that the number of no-submissions for each is varied. Act as instructor again and grade their submissions. Make sure the grades are spread over a reasonable range and pay attention to the ups and downs of one particular user. Take note of each user's trend so that comparison is easier for later purposes. After all of this is done, make sure to change the end date of each assignment to a past date. Unreleased grades are not considered for the grade-related conditions.
**Testing is first done case by case.**
4. Add a grace day usage condition to the course. Visit AutoPopulated/metrics/refresh_watch_list_instances. See if the returned instances match expectation. Tweak the condition to see if refreshed instances reflect the modification.
5. Add a grade drop condition. Visit AutoPopulated/metrics/refresh_watch_list_instances. See if the returned instances match expectation. Tweak the condition to see if refreshed instances reflect the modification.
6. Add a no submissions condition. Visit AutoPopulated/metrics/refresh_watch_list_instances. See if the returned instances match expectation. Tweak the condition to see if refreshed instances reflect the modification.
7. Add a low grades condition. Visit AutoPopulated/metrics/refresh_watch_list_instances. See if the returned instances match expectation. Tweak the condition to see if refreshed instances reflect the modification.
**Testing is then done in a hybrid manner.**
8. Update current conditions such that they are a hybrid of some of the four types. See if the refreshed instances match expectation.
**Lastly**
9. Refresh twice in a row. There shouldn't be any change in the records returned by the endpoint.
10. It should also be the case that none of the returned instances should include instructors' information.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [ ] My change requires a change to the documentation, which is located at [Autolab Docs](https://github.com/autolab/docs)
- [ ] I have updated the documentation accordingly, included in this PR

## Other issues / help required
<!--- Do you have any other relevant issues / questions? --->
<!--- For example, if you require assistance for a certain part of your PR --->
<!--- or are facing specific issues while creating the pull request that you would like to highlight --->

If unsure, feel free to submit first and we'll help you along.

victorhuangwq commented on app/models/watchlist_instance.rb:
```diff
@@ -22,6 +22,38 @@ def self.get_num_new_instance_for_course(course_name)
+
+    # check whether the watchlist instances are up-to-date
+    current_instances_condition_ids = (current_instances.map { |inst| inst.risk_condition_id }).uniq.to_set
+    current_condition_ids = (current_conditions.map { |c| c.id }).uniq.to_set # should be unique by design but add the suffix to ensure
```
This condition check for to reduce calculation actually causes a problem. Because if the conditions are kept the same, but the assignment deadlines etc are changed, this will actually prevent the instructor from updating the watchlist.

victorhuangwq replied:
Removing this check solved my issue

victorhuangwq commented on app/models/watchlist_instance.rb:
```diff
@@ -83,4 +115,176 @@ def resolve_watchlist_instance
+          asmt_arrs = categories.map { |category| course.assessments_with_category(category).ordered }
+          asmt_arrs.select! { |asmts| asmts.count >= consecutive_counts}
+          for cud in course_user_data
+            new_instance = self.add_new_instance_for_cud_grade_drop(course, condition.id, cud, asmt_arrs, consecutive_counts, percentage_drop)
```
should typecheck and convert consecutive counts to integers, else the comparison will fail later

victorhuangwq replied:
percentage_drop should be float

victorhuangwq commented on app/models/watchlist_instance.rb:
```diff
@@ -83,4 +115,176 @@ def resolve_watchlist_instance
+          no_submissions_threshold = condition.parameters[:no_submissions_threshold]
+
+          for cud in course_user_data
+            new_instance = self.add_new_instance_for_cud_no_submissions(course, condition.id, cud, no_submissions_threshold)
```
should typecheck and convert no_submissions_threshold to integer, else the comparison will fail later

victorhuangwq commented on app/models/watchlist_instance.rb:
```diff
@@ -83,4 +115,176 @@ def resolve_watchlist_instance
+          count_threshold = condition.parameters[:count_threshold]
+
+          for cud in course_user_data
+            new_instance = self.add_new_instance_for_cud_low_grades(course, condition.id, cud, grade_threshold, count_threshold)
```
should typecheck and convert grade_threshold, count_threshold to integer, else the comparison will fail later

victorhuangwq replied:
grade_threshold should be float

victorhuangwq commented on app/models/watchlist_instance.rb:
```diff
@@ -83,4 +115,176 @@ def resolve_watchlist_instance
+            asmts_before_date = asmts.where("updated_at < ?", date)
+            latest_asmt = asmts_before_date.last
+            next if latest_asmt.nil?
+            new_instance = self.add_new_instance_for_cud_grace_day_usage(course, condition.id, cud, asmts_before_date, grace_day_threshold)
```
should typecheck and convert grace_day_threshold to integers, else the comparison will fail later

victorhuangwq commented on app/models/watchlist_instance.rb:
```diff
@@ -83,4 +115,176 @@ def resolve_watchlist_instance
+          
+          categories = course.assessment_categories
+          asmt_arrs = categories.map { |category| course.assessments_with_category(category).ordered }
+          asmt_arrs.select! { |asmts| asmts.count >= consecutive_counts}
```
here as well, need to do a typecasting for consecutive_counts to ensure that it is an int

victorhuangwq commented:
Testing Steps:
1. Create a new course with 3 users
1. Create 3 assignments
1. Tested No Submissions, check to see if  appropriate results based on threshold. Changed Assignment deadlines and tested as well
1. Tested Low Grades. Check to see if appropriate results are obtained based on threshold.
1. Tested Grade Drop. 
1. Tested Grace Days Used.

Below are the issues that I have found

**Functional Issues Found**
- [ ] No Submissions Considered Assignments that are still ongoing. Given three assignments, 1, 2 & 3, and the watchlist is looking out for 2 no submissions. The student did not submit assignment 1. At the same time, the student did not submit assignment 3 yet, because it is still ongoing. 
**Expected**: Student should not be counted 
**Actual** Student considered as two no submissions. 
**Suggestion** Do no consider assignments that are still ongoing 
- [ ] Low Grades Assignment Considered Assignments that Grades have not been released. Given an assignments, and the watchlist is looking out for low grades of less than 50. The  student obtained 100 points, but the grades has not been released.
**Expected**: Student should not have considered as having low grades, because the assignment's grades has not been released
**Actual** Student considered as having low grades
**Suggestion** Do no consider grades of assignments that has not been released

victorhuangwq reviewed (changes requested):
The main overall issues to consider:
- The types of the parameters coming from `condition.parameters` are likely to be uncertain (can be string or integer). So they need to be typechecked and typecasted accordingly.
- Ensuring that the grading deadline and grade has been released for low grades, as well as when assignment has been submitted for no_submission

xinyis991105 commented:
> Testing Steps:
> 
>     1. Create a new course with 3 users
> 
>     2. Create 3 assignments
> 
>     3. Tested No Submissions, check to see if  appropriate results based on threshold. Changed Assignment deadlines and tested as well
> 
>     4. Tested Low Grades. Check to see if appropriate results are obtained based on threshold.
> 
>     5. Tested Grade Drop.
> 
>     6. Tested Grace Days Used.
> 
> 
> Below are the issues that I have found
> 
> **Functional Issues Found**
> 
>     * [x]  No Submissions Considered Assignments that are still ongoing. Given three assignments, 1, 2 & 3, and the watchlist is looking out for 2 no submissions. The student did not submit assignment 1. At the same time, the student did not submit assignment 3 yet, because it is still ongoing.
>       **Expected**: Student should not be counted
>       **Actual** Student considered as two no submissions.
>       **Suggestion** Do no consider assignments that are still ongoing
> 
>     * [ ]  Low Grades Assignment Considered Assignments that Grades have not been released. Given an assignments, and the watchlist is looking out for low grades of less than 50. The  student obtained 100 points, but the grades has not been released.
>       **Expected**: Student should not have considered as having low grades, because the assignment's grades has not been released
>       **Actual** Student considered as having low grades
>       **Suggestion** Do no consider grades of assignments that has not been released

There is actually a problem with the low grades situation. I managed to get the check for whether the final score has been released working for students who submitted something. But for those who didn't submit, I believe Autolab defaults to 0.0 as a final score for both when the student or the instructor tries to view the final score. In this case, suppose the grading deadline has already passed but a student didn't make any submission at all, do we just take it a 0 or would we rather check whether the same assessment's score has been released for other students? The latter might take a while a figure out though.

victorhuangwq commented:
So the situation is that :
1. the student did not submit
1. grades are released, they get a 0 for their score
1. they are counted as having a low grade 

I was wondering if you could just do a quick check if there is a latest_submission_id in assessment_user_datum?

xinyis991105 commented:
You mean when a student didn't submit, we don't count that as low grades? That could also work. In fact, that's so much simpler, I'll do a latest submission check as a solution.

xinyis991105 commented:
New commit patches:
1. Resolved all the type casting issues. (See resolved conversation and relevant code.)
2. Added grading deadline check, latest submission check, and score released check for :grade_drop and :low_grades conditions. Refreshing watchlist instances this time won't consider any assessment user datum that falls into one of the three buckets aforementioned.
3. Removed check for change of current risk conditions at the beginning of refreshing.
4. For grace day usage check, instead of including assessment user data whose updated_at is before the specified date in the parameter condition, end_at date is used. It makes more sense since normally, students won't be allowed to submit after the end date.

victorhuangwq commented:
In addition to the previous tests, the following test was done
1. Checking if no submission condition considers ongoing assessment by creating a new assessment and calling the endpoint
1. No submission is not counted for grade drop or low grades

Everything seems okay!

victorhuangwq reviewed (approved):
LGTM
